dali/pipeline/data/buffer.h

```diff
@@ -243,80 +157,88 @@ class Buffer {
   // Helper function for cleaning up data storage. This unfortunately
   // has to be public so that we can bind it into the deleter of our
   // shared pointers
-  void DeleterHelper(void *ptr, TypeInfo type, Index size) {
+  void DeleterHelper(void *ptr, Index size) {
     // change to correct device for deletion
     // Note: Can't use device guard due to potentially not GPUBackend.
     int current_device = 0;
     if (std::is_same<Backend, GPUBackend>::value) {
-      CUDA_CALL(cudaGetDevice(&current_device));
+      auto err = cudaGetDevice(&current_device);
+
+      // It's possible that the CUDA driver is unloading / has unloaded
+      // before we get to deleting all buffers. In that case we catch the appropriate
+      // error code and simply return.
+      if (err == cudaErrorCudartUnloading) return;
+
       CUDA_CALL(cudaSetDevice(device_));
     }
-    type.template Destruct<Backend>(ptr, size);
-    Backend::Delete(ptr, size*type.size(), pinned_);
+    if (ptr) {
+      // Only deallocate, underlying data freed elsewhere
+      Backend::Delete(ptr, size, pinned_);
+    }
 
     // reset to original calling device for consistency
     if (std::is_same<Backend, GPUBackend>::value) {
       CUDA_CALL(cudaSetDevice(current_device));
     }
   }
 
+  /**
+   * @brief Resize the buffer.
+   * Returns whether the underlying allocation
+   * changed.
+   */
+  bool Resize(size_t new_size) {
+    return ResizeHelper(new_size);
+  }
+
+  void ShareData(void *ptr, size_t bytes, int device_id) {
+    data_.reset(ptr, [](void *) {});
+    num_bytes_ = bytes;
+    size_ = 0;
```

Comment:
Why 0?

Reply:
It does not really matter what it is set to if it shares data, but I would agree that size should be set to bytes and num_bytes_ should be 0 instead (to make it clearer that this buffer does not actually have any capacity).